compact_str/src/repr2/mod.rs

```diff
@@ -0,0 +1,856 @@
+            unsafe { mem::transmute(heap) }
+        }
+
+        #[cold]
+        fn empty() -> Repr {
+            EMPTY
+        }
+
+        if cap.is_heap() {
+            // We only hit this case if the provided String is > 16MB and we're on a 32-bit arch. We
+            // expect it to be unlikely, thus we hint that to the compiler
+            capacity_on_heap(s)
+        } else if og_cap == 0 {
+            // We don't expect converting from an empty String often, so we make this code path cold
+            empty()
+        } else {
+            let mut s = mem::ManuallyDrop::new(s.into_bytes());
+            let len = s.len();
+            let raw_ptr = s.as_mut_ptr();
+
+            let ptr = ptr::NonNull::new(raw_ptr).expect("string with capacity has null ptr?");
+            let heap = HeapBuffer { ptr, len, cap };
+
+            // SAFETY: `BoxString` and `Repr` are the same size
+            unsafe { mem::transmute(heap) }
+        }
+    }
+
+    /// Converts a [`Repr`] into a [`String`], in `O(1)` time, if possible
+    #[inline]
+    pub fn into_string(self) -> String {
+        let last_byte = self.last_byte();
+
+        #[cold]
+        fn into_string_heap(this: HeapBuffer) -> String {
+            // SAFETY: We know pointer is valid for `length` bytes
+            let slice = unsafe { core::slice::from_raw_parts(this.ptr.as_ptr(), this.len) };
+            // SAFETY: A `Repr` contains valid UTF-8
+            let s = unsafe { core::str::from_utf8_unchecked(slice) };
+
+            String::from(s)
+        }
+
+        if last_byte == HEAP_MASK {
+            // SAFTEY: this is only ever called if we're heap allocated
+            let heap_buffer: HeapBuffer = unsafe { mem::transmute(self) };
+
+            if heap_buffer.cap.is_heap() {
+                // We don't expect capacity to be on the heap often, so we mark it as cold
+                into_string_heap(heap_buffer)
+            } else {
+                // Wrap the BoxString in a ManuallyDrop so the underlying buffer doesn't get freed
+                let this = mem::ManuallyDrop::new(heap_buffer);
+
+                // SAFETY: We checked above to make sure capacity is valid
+                let cap = unsafe { this.cap.as_usize() };
+
+                // SAFETY:
+                // * The memory in `ptr` was previously allocated by the same allocator the standard
+                //   library uses, with a required alignment of exactly 1.
+                // * `length` is less than or equal to capacity, due to internal invaraints.
+                // * `capacity` is correctly maintained internally.
+                // * `BoxString` only ever contains valid UTF-8.
+                unsafe { String::from_raw_parts(this.ptr.as_ptr(), this.len, cap) }
+            }
+        } else {
+            let pointer = &self as *const _ as *const u8;
+            let length = core::cmp::min((last_byte.wrapping_sub(LENGTH_MASK)) as usize, MAX_SIZE);
```

Comment:
I've seen this a couple of time, maybe it can be extracted as a separate function?

Reply:
Not a bad idea, I'll do this in a follow up PR